File: ncPersistencia/pParametro.vb
Imports ncDados.nsParametro
Imports ncComum.nsAcessoBD
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao
Imports ncDados.nsdParametroEstoque
Imports ncDados
Imports System.Security.Cryptography

Namespace nsParametro

    Public Class pParametro

        Public Function Listar() As ColecaoParametro

            Dim retorno As ColecaoParametro
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParametro
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " Select cid, descricao, valor From Parametros "

                ds = acessoBanco.ExecutarDS(comandoSQL)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParametro

                            For Each row In dt.Rows
                                item = New dParametro

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.descricao = cFuncoes.RetornarTexto(row("descricao"))
                                item.valor = cFuncoes.RetornarTexto(row("valor"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing

                Throw New ExcecaoNascomercio("Erro em Listar Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Listar = retorno

        End Function

        Public Function Consultar(ByVal dados As dParametro) As ColecaoParametro

            Dim retorno As ColecaoParametro
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParametro
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String

            Try

                acessoBanco = New cAcessoBD


                sqlSelect = " Select cid, descricao, valor "
                sqlWhere = String.Empty
                sqlFrom = " From Parametros "

                '-- cid
                sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cid, "cid")

                '-- descricao
                sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.descricao, "descricao")

                '-- valor
                sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.valor, "valor")

                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParametro

                            For Each row In dt.Rows
                                item = New dParametro

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.descricao = cFuncoes.RetornarTexto(row("descricao"))
                                item.valor = cFuncoes.RetornarTexto(row("valor"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch nex As ExcecaoNascomercio

                Throw

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Consultar = retorno

        End Function

        Public Function ConsultarEstoque(ByVal dados As dParametroEstoque) As ColecaoParametroEstoque

            Dim retorno As ColecaoParametroEstoque
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dEstoque
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String

            Try

                acessoBanco = New cAcessoBD


                sqlSelect = " SELECT e.fabricante, e.cid, p.descricao, e.referencia, e.item, e.valorCompra, e.valorVenda, e.valor, c.nome as cor"

                sqlWhere = String.Empty
                sqlFrom = " FROM nascomercio.produtos as p "
                sqlFrom += "  INNER JOIN nascomercio.v_estoque as e ON e.cid = p.cid "
                sqlFrom += "  INNER Join nascomercio.cor as c ON c.cid = p.cor_cid "

                If dados.valor = 1 Then
                    sqlWhere = sqlWhere + " e.valor > 0"
                Else
                    sqlWhere = sqlWhere + " e.valor = 0"
                End If

                If dados.cidGrupo > 0 Then
                    sqlWhere = sqlWhere + " AND p.grupo_cid = " + dados.cidGrupo.ToString
                End If

                If dados.cidFornecedor > 0 Then
                    sqlWhere = sqlWhere + " AND e.fornecedor_cid = " + dados.cidFornecedor.ToString
                End If

                If dados.cidFabricante > 0 Then
                    sqlWhere = sqlWhere + " AND e.fabricante_cid = " + dados.cidFabricante.ToString
                End If

                If Not String.IsNullOrEmpty(dados.descricao) Then
                    sqlWhere = sqlWhere + "AND descricao = '" + dados.descricao + "'"
                End If


                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere + " order by p.codigo, e.item")

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParametroEstoque

                            For Each row In dt.Rows
                                item = New dEstoque
                                item.Fabricante = cFuncoes.RetornarTexto(row("fabricante"))
                                item.CID = cFuncoes.RetornarTexto(row("cid"))
                                item.Descricao = cFuncoes.RetornarTexto(row("descricao"))
                                item.Referencia = cFuncoes.RetornarTexto(row("referencia"))
                                item.Item = cFuncoes.RetornarTexto(row("item"))
                                item.ValorCompra = cFuncoes.RetornarDecimal(row("valorCompra"))
                                item.ValorVenda = cFuncoes.RetornarDecimal(row("valorVenda"))
                                item.Descricao = cFuncoes.RetornarTexto(row("descricao"))
                                item.Valor = cFuncoes.RetornarDecimal(row("valor"))
                                item.Cor = cFuncoes.RetornarTexto(row("cor"))
                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch nex As ExcecaoNascomercio

                Throw

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            ConsultarEstoque = retorno

        End Function

        Public Function Incluir(ByVal dados As dParametro) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " INSERT INTO " & _
                    " Parametros ( descricao, valor ) " & _
                    " VALUES (" & _
                    cFuncoes.PersistirTexto(dados.descricao) & "," & _
                    cFuncoes.PersistirTexto(dados.valor) & ")"

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Incluir = retorno

        End Function

        Public Function Alterar(ByVal dados As dParametro) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " UPDATE Parametros SET " & _
                    " descricao = " & cFuncoes.PersistirTexto(dados.descricao) & "," & _
                    " valor = " & cFuncoes.PersistirTexto(dados.valor) & _
                    " WHERE " & _
                    " cid = " & dados.cid.ToString()

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Alterar = retorno

        End Function

        Public Function Excluir(ByVal dados As dParametro) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD

                comandoSQL = " DELETE FROM Parametros " & _
                    " WHERE cid = " & dados.cid.ToString()

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Parametro [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Excluir = retorno

        End Function

    End Class

End Namespace

File: ncRepositorios/pParametro.cs
using System;
using System.Collections.Generic;
using Dapper;
using MySql.Data.MySqlClient;
using ncNComum.nsExcecao;
using nsParametro;
using nsdParametroEstoque;

namespace ncPersistencia.nsParametro
{
    public class pParametro : RepositorioBase, IpParametro
    {
        public ColecaoParametro Listar()
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var lista = conn.Query<dParametro>("SELECT cid, descricao, valor FROM Parametros").AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParametro();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Listar Parametro [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParametro Consultar(dParametro dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (dados.cid != null && dados.cid != 0) { conditions.Add("cid=@cid"); p.Add("cid", dados.cid); }
                    if (!string.IsNullOrEmpty(dados.descricao)) { conditions.Add("descricao=@descricao"); p.Add("descricao", dados.descricao); }
                    if (!string.IsNullOrEmpty(dados.valor)) { conditions.Add("valor=@valor"); p.Add("valor", dados.valor); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT cid, descricao, valor FROM Parametros {where}";
                    var lista = conn.Query<dParametro>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParametro();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Parametro [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParametroEstoque ConsultarEstoque(dParametroEstoque dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    conditions.Add(dados.valor == "1" ? "e.valor > 0" : "e.valor = 0");
                    if (dados.cidGrupo > 0) { conditions.Add("p.grupo_cid = @grupo"); p.Add("grupo", dados.cidGrupo); }
                    if (dados.cidFornecedor > 0) { conditions.Add("e.fornecedor_cid = @fornecedor"); p.Add("fornecedor", dados.cidFornecedor); }
                    if (dados.cidFabricante > 0) { conditions.Add("e.fabricante_cid = @fabricante"); p.Add("fabricante", dados.cidFabricante); }
                    if (!string.IsNullOrEmpty(dados.descricao)) { conditions.Add("descricao = @descricao"); p.Add("descricao", dados.descricao); }
                    var where = "WHERE " + string.Join(" AND ", conditions);
                    var sql = $"SELECT e.fabricante AS Fabricante, e.cid AS CID, p.descricao AS Descricao, e.referencia AS Referencia, " +
                              $"e.item AS Item, e.valorCompra AS ValorCompra, e.valorVenda AS ValorVenda, e.valor AS Valor, c.nome AS Cor " +
                              $"FROM nascomercio.produtos AS p " +
                              $"INNER JOIN nascomercio.v_estoque AS e ON e.cid = p.cid " +
                              $"INNER JOIN nascomercio.cor AS c ON c.cid = p.cor_cid " +
                              $"{where} ORDER BY p.codigo, e.item";
                    var lista = conn.Query<dEstoque>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParametroEstoque();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Parametro [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Incluir(dParametro dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    conn.Execute("INSERT INTO Parametros (descricao, valor) VALUES (@descricao, @valor)",
                        new { descricao = dados.descricao, valor = dados.valor });
                    return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Incluir Parametro [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Alterar(dParametro dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("UPDATE Parametros SET descricao=@descricao, valor=@valor WHERE cid=@cid",
                        new { descricao = dados.descricao, valor = dados.valor, cid = dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Alterar Parametro [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Excluir(dParametro dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("DELETE FROM Parametros WHERE cid=@cid", new { cid = dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Excluir Parametro [" + ToString() + "] - " + ex.Message);
            }
        }
    }
}
